### PR TITLE
Receive: Make head series limiting config per tenant

### PR DESCRIPTION
**Pull request by saswatamcode**
This PR adds head/active series limiting configuration to the new per tenant limiting config added in #5565.
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

- Remove flags and fields passed via receive Handler options
- Make naming of active series limiting interfaces/structs consistent to "head series" and move to a separate file
- Add options to limiting config file
- Update e2e test to use new config
- Update docs

## Verification

All receive e2e tests pass locally and have tested different config scenarios.

### CODE REVIEW

File: docs/components/receive.md [hunk 1]
```diff
@@ -168,11 +173,15 @@ Thanos Receive, in Router or RouterIngestor mode, supports limiting tenant activ
...
 Every Receive Router/RouterIngestor node, queries meta-monitoring for active series of all tenants, every 15 seconds, and caches the results in a map. This cached result is used to limit all incoming remote write requests.
 
-To use the feature, one should specify the following (hidden) flags:
-- `--receive.tenant-limits.max-head-series`: Specifies the total number of active (head) series for any tenant, across all replicas (including data replication), allowed by Thanos Receive.
-- `--receive.tenant-limits.meta-monitoring-url`: Specifies Prometheus Query API compatible meta-monitoring endpoint.
-- `--receive.tenant-limits.meta-monitoring-query`: Optional flag to specify PromQL query to execute against meta-monitoring.
-- `--receive.tenant-limits.meta-monitoring-client`: Optional YAML file/string specifying HTTP client config for meta-monitoring.
+To use the feature, one should specify the following limiting config options:
+
+Under `global`,
```
Comment (douglascamata) [inline docs/components/receive.md:178]:
```suggestion
Under `global`:
```

File: docs/components/receive.md [hunk 2]
```diff
@@ -168,11 +173,15 @@ Thanos Receive, in Router or RouterIngestor mode, supports limiting tenant activ
...
+- `meta_monitoring_url`: Specifies Prometheus Query API compatible meta-monitoring endpoint.
+- `meta_monitoring_limit_query`: Option to specify PromQL query to execute against meta-monitoring. If not specified it is set to `sum(prometheus_tsdb_head_series) by (tenant)` by default.
+- `meta_monitoring_http_client`: Optional YAML field specifying HTTP client config for meta-monitoring.
+
+Under `default` and per `tenant`,
```
Comment (douglascamata) [inline docs/components/receive.md:183]:
```suggestion
Under `default` and per `tenant`:
```

File: pkg/receive/head_series_limiter.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+			}
+		}
+	}
+
+	return nil
+}
+
+// isUnderLimit ensures that the current number of active series for a tenant does not exceed given limit.
+// It does so in a best-effort way, i.e, in case meta-monitoring is unreachable, it does not impose limits.
+func (h *headSeriesLimit) isUnderLimit(tenant string, logger log.Logger) (bool, error) {
```
Comment (douglascamata) [inline pkg/receive/head_series_limiter.go:143]: Maybe the logger could be part of the `headSeriesLimit` struct?

File: pkg/receive/limiter_config.go [hunk 4]
```diff
@@ -21,37 +24,80 @@ func ParseRootLimitConfig(content []byte) (*RootLimitsConfig, error) {
...
-type writeLimitConfig struct {
+type WriteLimitConfig struct {
 	// RequestLimits holds the difficult per-request limits.
 	RequestLimits *requestLimitsConfig `yaml:"request"`
-	// HeadSeriesConfig *headSeriesLimiter `yaml:"head_series"`
+	// HeadSeriesLimit specifies the maximum number of head series allowed for a tenant.
+	HeadSeriesLimit *uint64 `yaml:"head_series_limit"`
+}
+
+// Utils for initialzing.
```
Comment (douglascamata) [inline pkg/receive/limiter_config.go:88]:
```suggestion
// Utils for initializing.
```

File: pkg/receive/request_limiter.go [hunk 5]
```diff
@@ -19,14 +19,20 @@ var unlimitedRequestLimitsConfig = newEmptyRequestLimitsConfig().
 	SetSeriesLimit(0).
 	SetSamplesLimit(0)
 
+type requestLimiter interface {
+	AllowSizeBytes(tenant string, contentLengthBytes int64) bool
+	AllowSeries(tenant string, amount int64) bool
+	AllowSamples(tenant string, amount int64) bool
+}
+
```
Comment (douglascamata) [inline pkg/receive/request_limiter.go:27]: Why was this moved far away from the place where it's used?

File: pkg/receive/request_limiter.go [hunk 6]
```diff
@@ -39,7 +45,9 @@ func newConfigRequestLimiter(reg prometheus.Registerer, writeLimits *writeLimits
 	tenantsLimits := writeLimits.TenantsLimits
 	tenantRequestLimits := make(map[string]*requestLimitsConfig)
 	for tenant, limitConfig := range tenantsLimits {
-		tenantRequestLimits[tenant] = limitConfig.RequestLimits.OverlayWith(defaultRequestLimits)
+		if limitConfig.RequestLimits != nil {
```
Comment (douglascamata) [inline pkg/receive/request_limiter.go:48]: Nice catch!

File: pkg/receive/head_series_limiter.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+func NewHeadSeriesLimit(w WriteLimitsConfig, registerer prometheus.Registerer, logger log.Logger) *headSeriesLimit {
+	limit := &headSeriesLimit{
+		metaMonitoringURL:   w.GlobalLimits.metaMonitoringURL,
+		metaMonitoringQuery: w.GlobalLimits.MetaMonitoringLimitQuery,
+		defaultLimit:        w.DefaultLimits.HeadSeriesLimit,
+		configuredTenantLimit: promauto.With(registerer).NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "thanos_receive_tenant_head_series_limit",
```
Comment (douglascamata) [inline pkg/receive/head_series_limiter.go:52]: I think you can skip the `tenant` in the name of this metric, given that it already has a tenant label.

File: pkg/receive/head_series_limiter.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+				Help: "The configured limit for active (head) series of tenants.",
+			}, []string{"tenant"},
+		),
+		configuredDefaultLimit: promauto.With(registerer).NewGauge(
```
Comment (douglascamata) [inline pkg/receive/head_series_limiter.go:56]: I think this metric is not necessary. You could follow the example of the request limiter and use `""` as value for the `"tenant"` label when the value is independent of the tenant.

**Review by douglascamata — CHANGES_REQUESTED** [review]
Great job overall! I put a few comments, but not many. 👍

File: pkg/receive/request_limiter.go (hunk 5 above)
Comment (saswatamcode) [reply pkg/receive/request_limiter.go:27]: I thought it might be better to have interface and implementation in the same file, but you're right, it might make it more confusing. Let me also move the `headSeriesLimiter` interface as well. 🙂

File: pkg/receive/head_series_limiter.go (hunk 8 above)
Comment (saswatamcode) [reply pkg/receive/head_series_limiter.go]: Removed and used empty tenant label.

**Review by douglascamata — APPROVED** [review]
LGTM, but would love to hear some feedback from the maintainers as well.

File: docs/components/receive.md [hunk 9]
```diff
@@ -101,34 +101,39 @@ The configuration file follows a few standards:
...
 
 From the example configuration below, it's understood that:
 
 1. This Receive instance has a max concurrency of 30.
-2. This Receive instance has some default request limits that apply of all tenants, **unless** a given tenant has their own limits (i.e. the `acme` tenant and partially for the `ajax` tenant).
-3. Tenant `acme` has no request limits.
-4. Tenant `ajax` has a request series limit of 50000 and samples limit of 500. Their request size bytes limit is inherited from the default, 1024 bytes.
+2. This Receive instance has head series limiting enabled as it has `meta_monitoring_.*` options in `global`.
+3. This Receive instance has some default request limits as well as head series limits that apply of all tenants, **unless** a given tenant has their own limits (i.e. the `acme` tenant and partially for the `ajax` tenant).
+4. Tenant `acme` has no request limits, but as a higher head_series limit.
```
Comment (matej-g) [inline docs/components/receive.md:113]:
```suggestion
4. Tenant `acme` has no request limits, but has a higher head_series limit.
```

File: docs/components/receive.md [hunk 10]
```diff
@@ -101,34 +101,39 @@ The configuration file follows a few standards:
...
 - `tenants`: the limits for a given tenant.
 
 From the example configuration below, it's understood that:
 
 1. This Receive instance has a max concurrency of 30.
-2. This Receive instance has some default request limits that apply of all tenants, **unless** a given tenant has their own limits (i.e. the `acme` tenant and partially for the `ajax` tenant).
-3. Tenant `acme` has no request limits.
-4. Tenant `ajax` has a request series limit of 50000 and samples limit of 500. Their request size bytes limit is inherited from the default, 1024 bytes.
+2. This Receive instance has head series limiting enabled as it has `meta_monitoring_.*` options in `global`.
+3. This Receive instance has some default request limits as well as head series limits that apply of all tenants, **unless** a given tenant has their own limits (i.e. the `acme` tenant and partially for the `ajax` tenant).
```
Comment (matej-g) [inline docs/components/receive.md:112]: This numbering makes me feel like we're talking about difference instances, how about something like `Receive instance with configuration below...`.

Alternatively, we could drop the description or just add in-line comments in the YAML excerpt below, since the behavior of default vs. tenant-specific limits is clearly described above.

File: pkg/receive/head_series_limiter.go [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+			},
+		),
+		logger: log.With(logger, "component", "receive-head-series-limiter"),
+	}
+
+	// Record default limit with empty tenant label.
+	limit.configuredTenantLimit.WithLabelValues("").Set(float64(limit.defaultLimit))
+
+	// Initialize map for configured limits of each tenant.
+	limit.limit = map[string]uint64{}
```
Comment (matej-g) [inline pkg/receive/head_series_limiter.go:70]: The `limit.limit` seems a bit confusing, would naming the parameter something along the lines `limitsPerTenant` make it clearer?

**Review by matej-g — COMMENTED** [review]
Stellar job as always @saswatamcode :sparkles:. I have few small-ish suggestions, mostly about organizing the code, otherwise this looks good!

File: pkg/httpconfig/http.go [hunk 12]
```diff
@@ -98,7 +98,7 @@ var defaultTransportConfig TransportConfig = TransportConfig{
 }
 
 func NewClientConfigFromYAML(cfg []byte) (*ClientConfig, error) {
-	conf := &ClientConfig{TransportConfig: defaultTransportConfig}
+	conf := &ClientConfig{TransportConfig: DefaultTransportConfig}
```
[the remark this answers is not in the document]
Comment (saswatamcode) [reply pkg/httpconfig/http.go:101]: Yeah was thinking about this, will implement!

File: test/e2e/e2ethanos/services.go [hunk 13]
```diff
@@ -505,11 +505,33 @@ func (r *ReceiveBuilder) Init() e2e.InstrumentedRunnable {
...
+		cfg := receive.RootLimitsConfig{
+			WriteLimits: receive.WriteLimitsConfig{
+				GlobalLimits: receive.GlobalLimitsConfig{
+					MetaMonitoringURL:        r.metamonitoring,
+					MetaMonitoringLimitQuery: r.metamonitoringQuery,
+				},
+				DefaultLimits: receive.DefaultLimitsConfig{
+					HeadSeriesLimit: uint64(r.limit),
+				},
+				// Can also add Per tenant limits.
```
Comment (bwplotka) [inline test/e2e/e2ethanos/services.go:517]: Do we want to merge this dead code?

**Review by bwplotka — APPROVED** [review]
Generally LGTM, mod @matej-g comments. Thanks!

File: test/e2e/e2ethanos/services.go (hunk 13 above)
Comment (saswatamcode) [reply test/e2e/e2ethanos/services.go:517]: I was keeping this to be able to test types of config and behaviour fully, but can remove! 🙂

**Comment by saswatamcode** [comment]
Have implemented all suggestions. Also, initialized `limiter` in `cmd/receive.go` and passed it down to `webHandler` as suggested by @matej-g . 🙂

**Comment by matej-g** [comment]
Seems like test failure is related :monocle_face:

```
--- FAIL: TestReceiveQuorum (0.00s)
    --- FAIL: TestReceiveQuorum/size_1_success (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xea9352]

goroutine 64 [running]:
testing.tRunner.func1.2({0xfb8c40, 0x1b41ff0})
	/usr/local/go/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1392 +0x39f
panic({0xfb8c40, 0x1b41ff0})
	/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/thanos-io/thanos/pkg/receive.(*Handler).receiveHTTP.func1({0x140fd58?, 0xc0000440b0?})
	/home/circleci/project/pkg/receive/handler.go:411 +0x52
github.com/thanos-io/thanos/pkg/tracing.DoInSpan({0x140fd58?, 0xc0000440b0?}, {0x1121e8c?, 0x2?}, 0xc00028fcc0, {0x0?, 0x0?, 0xa?})
	/home/circleci/project/pkg/tracing/tracing.go:95 +0xa3
github.com/thanos-io/thanos/pkg/receive.(*Handler).receiveHTTP(0xc0001942a0, {0x140f148, 0xc00022a6c0}, 0xc000152400)
	/home/circleci/project/pkg/receive/handler.go:410 +0x2b0
github.com/thanos-io/thanos/pkg/receive.makeRequest(0xc0001942a0, {0x1105819, 0x4}, 0x1?)
	/home/circleci/project/pkg/receive/handler_test.go:1210 +0x25e
github.com/thanos-io/thanos/pkg/receive.TestReceiveQuorum.func4(0xc0002621a0)
	/home/circleci/project/pkg/receive/handler_test.go:687 +0xd2
testing.tRunner(0xc0002621a0, 0xc00023c7c0)
	/usr/local/go/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1486 +0x35f
FAIL	github.com/thanos-io/thanos/pkg/receive	0.053s
```

File: pkg/receive/limiter_config.go [hunk 14]
```diff
@@ -21,37 +24,84 @@ func ParseRootLimitConfig(content []byte) (*RootLimitsConfig, error) {
...
+type WriteLimitsConfig struct {
 	// GlobalLimits are limits that are shared across all tenants.
-	GlobalLimits globalLimitsConfig `yaml:"global"`
+	GlobalLimits GlobalLimitsConfig `yaml:"global"`
 	// DefaultLimits are the default limits for tenants without specified limits.
-	DefaultLimits defaultLimitsConfig `yaml:"default"`
+	DefaultLimits DefaultLimitsConfig `yaml:"default"`
 	// TenantsLimits are the limits per tenant.
-	TenantsLimits tenantsWriteLimitsConfig `yaml:"tenants"`
+	TenantsLimits TenantsWriteLimitsConfig `yaml:"tenants"`
```
Comment (matej-g) [inline pkg/receive/limiter_config.go:60]: Is there a reason to export these methods? Apart from being used in E2E tests, looks like this is not really necessary. Maybe we could configure the E2E tests with YAML? (Might even be beneficial, since we can get config parsing in the E2E test for "free").

Comment (matej-g) [reply pkg/receive/limiter_config.go:60]: Clarified F2F, @saswatamcode is right that we do this in Thanos by standard and it eases configuring things when e.g. testing, demoing etc., resolving.

**Review by matej-g — DISMISSED** [review]

**Review by bwplotka — APPROVED** [review]
Thanks!

**Review by matej-g — APPROVED** [review]
Thanks for rebasing the PR! Looking good.
